### PR TITLE
Bots

### PR DESCRIPTION
**Pull request by ghost**
Can patrol assigned points sequentially.
Each Bot has a field that represents an array of Target Points and can be edited from UE4 Editor. If the array is not empty Bot will visit the location of each Target Point sequentially and stand there for the time period stated in Behavior Tree. 
Bot commits infinite number of visits.

### CODE REVIEW

File: Source/Prot/AI/BTTargetPointSelection.cpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+		{
+			PointIndex = 0;
+		}
+		BlackboardComp->SetValueAsInt("PointIndex", PointIndex);
+
+		//At this point, the task has been successfully completed
+		return EBTNodeResult::Succeeded;
+	}
+	return EBTNodeResult::Failed;
+}
```
Comment (UnoYakshi) [inline Source/Prot/AI/BTTargetPointSelection.cpp]: Add a newline.

File: Source/Prot/AI/BTTargetPointSelection.cpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+#include "BTTargetPointSelection.h"
+#include "BotAIC.h"
+#include "BehaviorTree/BlackboardComponent.h"
+#include "Engine/TargetPoint.h"
+
+EBTNodeResult::Type UBTTargetPointSelection::ExecuteTask(UBehaviorTreeComponent & OwnerComp, uint8 * NodeMemory)
+{
+	ABotAIC* AICon = Cast<ABotAIC>(OwnerComp.GetAIOwner());
+
```
Comment (UnoYakshi) [inline Source/Prot/AI/BTTargetPointSelection.cpp]: Abundant newline.

File: Source/Prot/AI/BTTargetPointSelection.cpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	if (AICon)
+	{
+		UBlackboardComponent* BlackboardComp = AICon->GetBlackboardComp();
+
+		//Update the next location in the Blackboard so the bot can move to the next Blackboard value
```
Comment (UnoYakshi) [inline Source/Prot/AI/BTTargetPointSelection.cpp]: What's `next Blackboard value`?

File: Source/Prot/AI/BTTargetPointSelection.h [hunk 4]
```diff
@@ -0,0 +1,18 @@
+// Fill out your copyright notice in the Description page of Project Settings.
+
+#pragma once
+
+#include "CoreMinimal.h"
+#include "BehaviorTree/BTTaskNode.h"
+#include "BTTargetPointSelection.generated.h"
+
+/**
+ *
```
Comment (UnoYakshi) [inline Source/Prot/AI/BTTargetPointSelection.h]: Add a description.

File: Source/Prot/AI/BotAIC.cpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+#include "Kismet/GameplayStatics.h"
+#include "Engine/TargetPoint.h"
+
+ABotAIC::ABotAIC()
+{
+	//Initialize BehaviorTreeComponent, BlackboardComponent
+
+	BehaviorComp = CreateDefaultSubobject<UBehaviorTreeComponent>(TEXT("BehaviorComp"));
+
+	BlackboardComp = CreateDefaultSubobject<UBlackboardComponent>(TEXT("BlackboardComp"));
```
Comment (UnoYakshi) [inline Source/Prot/AI/BotAIC.cpp]: Abundant newlines.

File: Source/Prot/AI/BotAIC.cpp [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+void ABotAIC::Possess(APawn* Pawn)
+{
+	Super::Possess(Pawn);
+
+	//Get the possessed Pawn and check if it's Bot
+	ABot* Bot = Cast<ABot>(Pawn);
+
```
Comment (UnoYakshi) [inline Source/Prot/AI/BotAIC.cpp]: Abundant newline.

File: Source/Prot/AI/BotAIC.cpp [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+	if (Bot)
+	{
+		//If the blackboard is valid initialize the blackboard for the corresponding behavior tree
+		if (Bot->BehaviorTree->BlackboardAsset)
+		{
+			BlackboardComp->InitializeBlackboard(*(Bot->BehaviorTree->BlackboardAsset));
+		}
+
+		BlackboardComp->SetValueAsInt("PointIndex", 0);
```
Comment (UnoYakshi) [inline Source/Prot/AI/BotAIC.cpp]: What if `BlackBoard` won't be initialized?

File: Source/Prot/AI/BotAIC.cpp [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+		return Bot->PatrolPoints[index];
+	}
+	return nullptr;
+}
+
+//Gets number of Target Points in Bot's array
+int ABotAIC::GetTargetPointsNumber()
+{
+	ABot* Bot = Cast<ABot>(GetPawn());
+
```
Comment (UnoYakshi) [inline Source/Prot/AI/BotAIC.cpp]: Abundant newline.

File: Source/Prot/AI/BotAIC.cpp [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+	if (Bot)
+	{
+		return Bot->PatrolPoints.Num();
+	}
+	return 0;
+}
```
Comment (UnoYakshi) [inline Source/Prot/AI/BotAIC.cpp]: Add a newline.

File: Source/Prot/AI/BotAIC.h [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+#pragma once
+
+#include "CoreMinimal.h"
+#include "AIController.h"
+#include "BehaviorTree/BehaviorTreeComponent.h"
+#include "Engine/TargetPoint.h"
+#include "BotAIC.generated.h"
+
+/**
+ *
```
Comment (UnoYakshi) [inline Source/Prot/AI/BotAIC.h]: Add a description.

File: Source/Prot/AI/BotAIC.h [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+ */
+UCLASS()
+class PROT_API ABotAIC : public AAIController
+{
+	GENERATED_BODY()
+
+	ABotAIC();
+
+	//Behavior tree component reference
+	UBehaviorTreeComponent* BehaviorComp;
```
Comment (UnoYakshi) [inline Source/Prot/AI/BotAIC.h]: Comment is abundant. We already know it's UBTComponent pointer by its signature.

File: Source/Prot/AI/BotAIC.h [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+	//Blackboard component reference
+	UBlackboardComponent* BlackboardComp;
```
Comment (UnoYakshi) [inline Source/Prot/AI/BotAIC.h]: Comment is abundant. We already know it's Blackboard pointer by its signature.

File: Source/Prot/AI/BotAIC.h [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+	Inside this function, we initialize the blackboard and start the behavior tree*/
+	virtual void Possess(APawn* Pawn) override;
+
+public:
+	FORCEINLINE UBlackboardComponent* GetBlackboardComp() const 
+	{
+		return BlackboardComp;
+	}
+
+	virtual ATargetPoint* GetTargetPointByIndex(int index) const;
```
Comment (UnoYakshi) [inline Source/Prot/AI/BotAIC.h:36]: Add a description.

File: Source/Prot/AI/BotAIC.h [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+	virtual int GetTargetPointsNumber();
```
Comment (UnoYakshi) [inline Source/Prot/AI/BotAIC.h:39]: Add a description.

File: Source/Prot/Bot.cpp [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+void ABot::BeginPlay()
+{
+	Super::BeginPlay();
+}
+
+// Called every frame
+void ABot::Tick(float DeltaTime)
+{
+	Super::Tick(DeltaTime);
+}
```
Comment (UnoYakshi) [inline Source/Prot/Bot.cpp]: Add a newline.

File: Source/Prot/Bot.cpp [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+	PrimaryActorTick.bCanEverTick = true;
+}
+
+// Called when the game starts or when spawned
+void ABot::BeginPlay()
+{
+	Super::BeginPlay();
+}
+
+// Called every frame
```
Comment (UnoYakshi) [inline Source/Prot/Bot.cpp]: Remove abundant comment.

File: Source/Prot/Bot.cpp [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+#include "GameFramework/CharacterMovementComponent.h"
+
+// Sets default values
+ABot::ABot()
+{
+ 	// Set this pawn to call Tick() every frame.  You can turn this off to improve performance if you don't need it.
+	PrimaryActorTick.bCanEverTick = true;
+}
+
+// Called when the game starts or when spawned
```
Comment (UnoYakshi) [inline Source/Prot/Bot.cpp]: Remove abundant comment.

File: Source/Prot/Bot.cpp [hunk 18]
```diff
@@ -0,0 +1,26 @@
+// Fill out your copyright notice in the Description page of Project Settings.
+
+#include "Bot.h"
+#include "NavigationSystem.h"
+#include "DrawDebugHelpers.h"
+#include "NavigationPath.h"
+#include "GameFramework/CharacterMovementComponent.h"
+
+// Sets default values
```
Comment (UnoYakshi) [inline Source/Prot/Bot.cpp]: Remove abundant comment.

File: Source/Prot/Bot.cpp [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+ABot::ABot()
+{
+ 	// Set this pawn to call Tick() every frame.  You can turn this off to improve performance if you don't need it.
```
Comment (UnoYakshi) [inline Source/Prot/Bot.cpp:11]: Remove abundant comment.

File: Source/Prot/Bot.h [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+#include "CoreMinimal.h"
+#include "GameFramework/Character.h"
+#include "NavigationPath.h"
+#include "GameFramework/PawnMovementComponent.h"
+#include "BehaviorTree/BehaviorTree.h"
+#include "Engine/TargetPoint.h"
+#include "Bot.generated.h"
+
+UCLASS()
+class PROT_API ABot : public ACharacter
```
Comment (UnoYakshi) [inline Source/Prot/Bot.h:15]: Add a description.

File: Source/Prot/Bot.h [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+{
+	GENERATED_BODY()
+
+public:
+	// Sets default values for this pawn's properties
```
Comment (UnoYakshi) [inline Source/Prot/Bot.h]: Remove abundant comment.

File: Source/Prot/Bot.h [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+	ABot();
+
+protected:
+	// Called when the game starts or when spawned
+	virtual void BeginPlay() override;
```
Comment (UnoYakshi) [inline Source/Prot/Bot.h:23]: Remove abundant comment.

File: Source/Prot/Bot.h [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+
+public:	
+	// Called every frame
+	virtual void Tick(float DeltaTime) override;
```
Comment (UnoYakshi) [inline Source/Prot/Bot.h:26]: Remove abundant comment.

File: Source/Prot/Bot.h [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+
+	UPROPERTY(EditAnywhere, BlueprintReadWrite, Category = "AI")
+	class UBehaviorTree* BehaviorTree;
```
Comment (UnoYakshi) [inline Source/Prot/Bot.h:30]: Add a description.

File: Source/Prot/Prot.Build.cs [hunk 25]
```diff
@@ -7,7 +7,26 @@ public class Prot : ModuleRules
...
+        // Uncomment if you are using online features
+        //PrivateDependencyModuleNames.Add("OnlineSubsystem");
+        //PrivateDependencyModuleNames.Add("OnlineSubsystemNull");
+        //if ((Target.Platform == UnrealTargetPlatform.Win32) || (Target.Platform == UnrealTargetPlatform.Win64) || (Target.Platform == UnrealTargetPlatform.Linux))
+        //{
+        //    DynamicallyLoadedModuleNames.Add("OnlineSubsystemSteam");
+        //}
+
+        // END STEAM INTEGRATION
+    }
```
Comment (UnoYakshi) [inline Source/Prot/Prot.Build.cs:15]: Remove the commented code as we don't need it now.

**Review by UnoYakshi — CHANGES_REQUESTED** [review]

**Comment by UnoYakshi** [comment]
Overall, the job is done. And we can call it a day.
Provide more meaningful descriptions both for PR and functionality. Pretend we don't know a thing on AI stuff, though, we do understand basic gameplay framework and C++.

File: Source/Prot/AI/BTTargetPointSelection.cpp [hunk 26]
```diff
@@ -0,0 +1,29 @@
+// Fill out your copyright notice in the Description page of Project Settings.
```
Comment (andrew-it) [inline Source/Prot/AI/BTTargetPointSelection.cpp:1]: Fill it or remove the comment

File: Source/Prot/AI/BTTargetPointSelection.cpp [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+
+#include "BTTargetPointSelection.h"
+#include "BotAIC.h"
+#include "BehaviorTree/BlackboardComponent.h"
+#include "Engine/TargetPoint.h"
+
+EBTNodeResult::Type UBTTargetPointSelection::ExecuteTask(UBehaviorTreeComponent & OwnerComp, uint8 * NodeMemory)
+{
+	ABotAIC* AICon = Cast<ABotAIC>(OwnerComp.GetAIOwner());
```
Comment (andrew-it) [inline Source/Prot/AI/BTTargetPointSelection.cpp]: Not clear what is AICon. Please, describe in comments or rename the var.

File: Source/Prot/AI/BTTargetPointSelection.h [hunk 28]
```diff
@@ -0,0 +1,16 @@
+// Fill out your copyright notice in the Description page of Project Settings.
```
Comment (andrew-it) [inline Source/Prot/AI/BTTargetPointSelection.h:1]: The same

File: Source/Prot/AI/BotAIC.cpp [hunk 29]
```diff
@@ -0,0 +1,64 @@
+// Fill out your copyright notice in the Description page of Project Settings.
```
Comment (andrew-it) [inline Source/Prot/AI/BotAIC.cpp:1]: The same

File: Source/Prot/AI/BotAIC.cpp [hunk 30]
```diff
@@ -0,0 +1,64 @@
...
+	BehaviorComp = CreateDefaultSubobject<UBehaviorTreeComponent>(TEXT("BehaviorComp"));
+	BlackboardComp = CreateDefaultSubobject<UBlackboardComponent>(TEXT("BlackboardComp"));
+}
+
+void ABotAIC::Possess(APawn* Pawn)
+{
+	Super::Possess(Pawn);
+
+	//Get the possessed Pawn and check if it's Bot
+	ABot* Bot = Cast<ABot>(Pawn);
```
Comment (andrew-it) [inline Source/Prot/AI/BotAIC.cpp:25]: Can a Pawn be of a different class? -> Is it safe way to cast?

File: Source/Prot/AI/BotAIC.h [hunk 31]
```diff
@@ -0,0 +1,36 @@
+// Fill out your copyright notice in the Description page of Project Settings.
```
Comment (andrew-it) [inline Source/Prot/AI/BotAIC.h:1]: the same

File: Source/Prot/AI/BotAIC.h (hunk 11 above)
Comment (andrew-it) [reply Source/Prot/AI/BotAIC.h]: where is the comment?

File: Source/Prot/AI/BotAIC.h (hunk 12 above)
Comment (andrew-it) [reply Source/Prot/AI/BotAIC.h]: where is the comment?

File: Source/Prot/Bot.cpp (hunk 19 above)
Comment (andrew-it) [reply Source/Prot/Bot.cpp:11]: @UnoYakshi, why?

File: Source/Prot/Prot.Build.cs [hunk 32]
```diff
@@ -1,4 +1,4 @@
-// Copyright 1998-2018 Epic Games, Inc. All Rights Reserved.
+// Fill out your copyright notice in the Description page of Project Settings.
```
Comment (andrew-it) [inline Source/Prot/Prot.Build.cs:1]: meow?

**Review by andrew-it — CHANGES_REQUESTED** [review]

File: Source/Prot/AI/BTTargetPointSelection.cpp (hunk 26 above)
Comment (ghost) [reply Source/Prot/AI/BTTargetPointSelection.cpp:1]: What exactly should I write here? @UnoYakshi

File: Source/Prot/AI/BotAIC.cpp (hunk 30 above)
Comment (ghost) [reply Source/Prot/AI/BotAIC.cpp:25]: @UnoYakshi what do you think?

File: Source/Prot/Prot.Build.cs (hunk 32 above)
Comment (UnoYakshi) [reply Source/Prot/Prot.Build.cs:1]: I will fix it in `dev` instead.

File: Source/Prot/AI/BotAIC.cpp (hunk 30 above)
Comment (UnoYakshi) [reply Source/Prot/AI/BotAIC.cpp:25]: Safe enough. We will probably inherit new bots from `ABot` in the future so it won't break anything.
As a summary, I don't see any issues here. It's a quite common way in UE4 framework.

File: Source/Prot/AI/BTTargetPointSelection.cpp (hunk 26 above)
Comment (UnoYakshi) [reply Source/Prot/AI/BTTargetPointSelection.cpp:1]: @oninbo, you shouldn't. I will set it up later, via Editor's Project Settings. It's just a meta-data for the project.

Comment (ghost) [reply Source/Prot/AI/BTTargetPointSelection.cpp:1]: 👌

File: Source/Prot/AI/BotAIC.h (hunk 11 above)
Comment (UnoYakshi) [reply Source/Prot/AI/BotAIC.h]: @oninbo, we still need a comment about what it is. But not a copy of the code. Also, there should be a newline between different items.

File: Source/Prot/AI/BotAIC.h (hunk 12 above)
Comment (UnoYakshi) [reply Source/Prot/AI/BotAIC.h]: @oninbo, same here. Add a meaningful comment here.

File: Source/Prot/Bot.h [hunk 33]
```diff
@@ -0,0 +1,36 @@
...
+	virtual void BeginPlay() override;
+
+public:	
+	virtual void Tick(float DeltaTime) override;
+
+	//The class of Behavior Tree that describes Bot's logic
+	UPROPERTY(EditAnywhere, BlueprintReadWrite, Category = "AI")
+	class UBehaviorTree* BehaviorTree;
+
+	// Points to patrol
```
Comment (UnoYakshi) [inline Source/Prot/Bot.h]: I'd suggest adding that a bot will patrol on them in the given order.

**Review by UnoYakshi — CHANGES_REQUESTED** [review]

File: Source/Prot/AI/BotAIC.cpp [hunk 34]
```diff
@@ -0,0 +1,64 @@
...
+			BehaviorComp->StartTree(*Bot->BehaviorTree);
+		}
+		else
+		{
+			UE_LOG(LogTemp, Error, TEXT("Blackboard was not valid"));
+		}
+	}
+}
+
+//Gets Target Point from Bot's array by index
```
Comment (UnoYakshi) [inline Source/Prot/AI/BotAIC.cpp]: Move the description to the header.

File: Source/Prot/AI/BotAIC.cpp [hunk 35]
```diff
@@ -0,0 +1,64 @@
...
+{
+	ABot* Bot = Cast<ABot>(GetPawn());
+	if (Bot && index < Bot->PatrolPoints.Num() && index >= 0)
+	{
+		return Bot->PatrolPoints[index];
+	}
+	return nullptr;
+}
+
+//Gets number of Target Points in Bot's array
```
Comment (UnoYakshi) [inline Source/Prot/AI/BotAIC.cpp]: Move the description to the header.

**Review by UnoYakshi — CHANGES_REQUESTED** [review]

**Review by UnoYakshi — APPROVED** [review]

**Pull request merged**